core/retries-api/src/main/java/software/amazon/awssdk/retriesapi/RetryStrategy.java

```diff
@@ -0,0 +1,208 @@
+     *
+     * <p>If the attempt was successful, callers MUST call {@link #recordSuccess}.
+     *
+     * @throws NullPointerException            if a required parameter is not specified
+     * @throws TokenAcquisitionFailedException if a token cannot be acquired
+     */
+    AcquireInitialTokenResponse acquireInitialToken(AcquireInitialTokenRequest request);
+
+    /**
+     * Invoked before each subsequent (non-first) request attempt.
+     *
+     * <p>Callers MUST wait for the {@code delay} returned by this call before making the next attempt. If the next attempt
+     * fails, callers MUST re-call {@link #refreshRetryToken} before attempting another retry. This call invalidates the provided
+     * token, and returns a new one. Callers MUST use the new token.
+     *
+     * <p>If the attempt was successful, callers MUST call {@link #recordSuccess}.
+     *
+     * @throws NullPointerException            if a required parameter is not specified
+     * @throws IllegalArgumentException        if the provided token was not issued by this strategy or the provided token was
+     *                                         already used for a previous refresh or success call.
+     * @throws TokenAcquisitionFailedException if a token cannot be acquired
+     */
+    RefreshRetryTokenResponse refreshRetryToken(RefreshRetryTokenRequest request);
+
+    /**
+     * Invoked after an attempt succeeds.
+     *
+     * @throws NullPointerException     if a required parameter is not specified
+     * @throws IllegalArgumentException if the provided token was not issued by this strategy or the provided token was already
+     *                                  used for a previous refresh or success call.
+     */
+    RecordSuccessResponse recordSuccess(RecordSuccessRequest request);
+
+    /**
+     * Create a new {@link Builder} with the current configuration.
+     *
+     * <p>This is useful for modifying the strategy's behavior, like conditions or max retries.
+     */
+    @Override
+    Builder toBuilder();
+
+    /**
+     * Builder to create immutable instances of {@link RetryStrategy}.
+     */
+    interface Builder extends CopyableBuilder<Builder, RetryStrategy> {
+        /**
+         * Configure the strategy to retry when the provided predicate returns true, given a failure exception.
+         */
+        Builder retryOnException(Predicate<Throwable> shouldRetry);
+
+        /**
+         * Configure the strategy to retry when a failure exception class is equal to the provided class.
+         */
+        default Builder retryOnException(Class<? extends Throwable> throwable) {
+            return retryOnException(t -> t.getClass() == throwable);
+        }
+
+        /**
+         * Configure the strategy to retry when a failure exception class is an instance of the provided class (includes
+         * subtypes).
+         */
+        default Builder retryOnExceptionInstanceOf(Class<? extends Throwable> throwable) {
```

Comment:
I guess the argument in favor would be that it will make some expressions easier but the same can be accomplish by calling this method for each. Since I'm not quite sure that this will add much I think it would be better to do that when and if the need araises.